readium/navigator/src/main/java/org/readium/r2/navigator/settings/Setting.kt

```diff
@@ -0,0 +1,354 @@
+        const val THEME = "theme"
+        const val TYPE_SCALE = "typeScale"
+        const val WORD_SPACING = "wordSpacing"
+
+        /**
+         * Keys of settings that are tied to a single publication and should not be shared between
+         * several publications.
+         */
+        @ExperimentalReadiumApi
+        val PUBLICATION_SETTINGS = listOf(
+            LANGUAGE,
+            READING_PROGRESSION,
+        ).toTypedArray()
+    }
+
+    /**
+     * JSON raw representation for the current value.
+     */
+    private val jsonValue: JsonElement = coder.encode(value)
+
+    /**
+     * Creates a copy of the [Setting] receiver, after replacing its value with the first valid
+     * value taken from the given [Preferences] objects, in order.
+     *
+     * Each preference is verified using the setting [validator].
+     */
+    fun copyFirstValidValueFrom(vararg candidates: Preferences?, fallback: Setting<V, E> = this): Setting<V, E> =
+        copy(
+            value = candidates
+                .mapNotNull { candidate -> candidate?.get(this) }
+                .firstNotNullOfOrNull(::validate)
+                ?: fallback.value
+        )
+
+    // We only want to use the setting key and value to test the structural equality, so we
+    // override the default implementation generated by the data class.
+
+    override fun equals(other: Any?): Boolean {
+        val otherSetting = (other as? Setting<*, *>) ?: return false
+        return otherSetting.key == key && otherSetting.jsonValue == jsonValue
+    }
+
+    override fun hashCode(): Int {
+        var result = key.hashCode()
+        result = 31 * result + jsonValue.hashCode()
+        return result
+    }
+}
+
+/**
+ * An arbitrary [Setting] without constraint except its type.
+ */
+@ExperimentalReadiumApi
+typealias ValueSetting<V> = Setting<V, Unit>
+
+/**
+ * Creates a new [ValueSetting] with the given [value].
+ */
+@ExperimentalReadiumApi
+inline fun <reified V> ValueSetting(
+    key: String,
+    value: V,
+    coder: SettingCoder<V> = SerializerSettingCoder(),
+    validator: SettingValidator<V> = IdentitySettingValidator(),
+    activator: SettingActivator = NullSettingActivator,
+) : ValueSetting<V> =
+    Setting(
+        key = key, value = value, extras = Unit,
+        coder = coder, validator = validator, activator = activator
+    )
+
+/**
+ * A boolean [Setting].
+ */
+@ExperimentalReadiumApi
+typealias ToggleSetting = ValueSetting<Boolean>
+
+/**
+ * Creates a new [ToggleSetting] with the given [value].
+ */
+@ExperimentalReadiumApi
+fun ToggleSetting(
+    key: String,
+    value: Boolean,
+    validator: SettingValidator<Boolean> = IdentitySettingValidator(),
+    activator: SettingActivator = NullSettingActivator,
+) : ToggleSetting =
+    Setting(
+        key = key, value = value, extras = Unit,
+        coder = SerializerSettingCoder(),
+        validator = validator, activator = activator
+    )
+
```

Comment:
Data classes don't support inheritance, but this will be solved with https://github.com/readium/kotlin-toolkit/pull/141/files/bce2d2431cb538676cf32862a6c2936dd2a47f8e#r962932147